src/aggregate/aggregate_request.c

```diff
@@ -1321,46 +1321,29 @@ static void SafeRedisKeyspaceAccessPipeline(AREQ *req) {
   ResultProcessor *upstream_is_unlcoker = NULL;
 
   ResultProcessor dummy_rp = {.upstream = req->qiter.endProc};
-  ResultProcessor *curr_rp = &dummy_rp;
+
   // Start from the end processor and iterate beackward until the next rp
   // is the last to access redis or its a pipeline breaker.
-
-  while (curr_rp != req->qiter.rootProc &&
-        !(curr_rp->upstream->flags & (RESULT_PROCESSOR_F_ACCESS_REDIS | RESULT_PROCESSOR_F_BREAKS_PIPELINE))) {
-    curr_rp = curr_rp->upstream;
-  }
-
-  // if we got the root proc, redis access in not needed, return.
-  if (curr_rp == req->qiter.rootProc) {
-    return;
-  }
-
-  // The upstream rp of the curr_rp is the last to access redis, or a pipline breaker
-  // we want to place the unlocker between curr_rp and its upstream.
-  upstream_is_unlcoker = curr_rp;
-
-  // The last to access redis might be also the first to access redis.
-  // We mark it to push the buffer-locker as its upstream.
-  curr_rp = curr_rp->upstream;
-
-  upstream_is_buffer_locker = curr_rp;
-
-  // Keep searching until we get to the root rp.
-  curr_rp = curr_rp->upstream;
-
-  while (curr_rp != req->qiter.rootProc) {
-    if (curr_rp->flags & RESULT_PROCESSOR_F_ACCESS_REDIS) {
+  for (ResultProcessor *curr_rp = &dummy_rp; curr_rp != req->qiter.rootProc; curr_rp = curr_rp->upstream) {
+    if (curr_rp->superClass == RESULT_PROCESSOR_C_ACCESS_REDIS) {
+      // If we found (another) RP that accesses redis, update the upstream_is_buffer_locker.
       upstream_is_buffer_locker = curr_rp;
+    } else if (curr_rp->superClass == RESULT_PROCESSOR_C_ACCUMULATOR && !upstream_is_buffer_locker) {
+      // If we found an accumulator and we didn't find any RP that accesses redis yet, reset the upstream_is_unlcoker.
+      upstream_is_unlcoker = NULL;
+    }
+    // If we didn't find a relevant RP for unlocker yet and the next RP access redis or is an aborter, set the upstream_is_unlcoker.
+    if (!upstream_is_unlcoker && (curr_rp->upstream->superClass == RESULT_PROCESSOR_C_ACCESS_REDIS ||
+                                  curr_rp->upstream->superClass == RESULT_PROCESSOR_C_ABORTER)) {
```

Comment:
validate scenarios without `RESULT_PROCESSOR_C_ACCESS_REDIS` and `RESULT_PROCESSOR_C_ABORTER`